proparse/src/main/java/org/prorefactor/proparse/antlr4/Lexer.java

```diff
@@ -0,0 +1,1117 @@
+        case '8':
+        case '9':
+        case 'a':
+        case 'b':
+        case 'c':
+        case 'd':
+        case 'e':
+        case 'f':
+        case 'g':
+        case 'h':
+        case 'i':
+        case 'j':
+        case 'k':
+        case 'l':
+        case 'm':
+        case 'n':
+        case 'o':
+        case 'p':
+        case 'q':
+        case 'r':
+        case 's':
+        case 't':
+        case 'u':
+        case 'v':
+        case 'w':
+        case 'x':
+        case 'y':
+        case 'z':
+        case '_':
+        case '-':
+        case '$':
+        case '#':
+        case '%':
+        case '&':
+        case '/':
+          // For tokens like ALT-* and CTRL-` :
+          // Emperically, I found that the following are the only other
+          // characters that can be put into a key label. Things
+          // like ALT-, must be put into quotes "ALT-,", because
+          // the comma has special meaning in 4gl code.
+          // Note also that extended characters can come after CTRL- or ALT-.
+          // ('!'|'@'|'^'|'*'|'+'|';'|'"'|'`')
+        case '!':
+        case '"':
+        case '*':
+        case '+':
+        case ';':
+        case '@':
+        case '^':
+        case '`':
+          append();
+          getChar();
+          break;
+        case '\\':
+        case '\'':
+          append();
+          getChar();
+          if (ttype == PreprocessorParser.ID)
+            ttype = PreprocessorParser.FILENAME;
+          break;
+        default:
+          if (currInt >= 128 && currInt <= 255) {
+            append();
+            getChar();
+            break;
+          } else {
+            break for_loop;
+          }
+      }
+    }
+    // See if it's a keyword
+    if (ttype == PreprocessorParser.ID)
+      ttype = NodeTypes.testLiteralsTable(currText.toString(), ttype);
+    return makeToken(ttype);
+  }
+
+  ProToken ampText() throws IOException {
+    LOGGER.trace("Entering ampText()");
+    for (;;) {
+      if (Character.isLetterOrDigit(currInt) || (currInt >= 128 && currInt <= 255)) {
+        append();
+        getChar();
+        continue;
+      }
+      switch (currChar) {
+        case '#':
+        case '$':
+        case '%':
+        case '&':
+        case '-':
+        case '_':
+          append();
+          getChar();
+          continue;
+      }
+      if (currChar == '/') {
+        // You can embed comments in (or at the end of) an &token.
+        // I've no idea why. See the regression test for bug#083.
+        preserveCurrent();
+        getChar();
+        if (currChar == '*') {
+          String s = currText.toString();
+          comment();
+          currText.replace(0, currText.length(), s);
+          preserveDrop();
+          continue;
+        }
+      }
+      break;
+    }
+    ProToken t = directive();
+    if (t != null)
+      return t;
+    return makeToken(PreprocessorParser.FILENAME);
+  }
+
+  ProToken directive() throws IOException {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "directive" is 21 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)